besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,369 @@
+    TrieLogHelper.validatePruneConfiguration(config);
+
+    final long layersToRetain = config.getUnstable().getBonsaiTrieLogRetentionThreshold();
+
+    final long chainHeight = blockchain.getChainHeadBlockNumber();
+
+    final long lastBlockNumberToRetainTrieLogsFor = chainHeight - layersToRetain;
+
+    if (!validPruneRequirements(blockchain, chainHeight, lastBlockNumberToRetainTrieLogsFor)) {
+      return;
+    }
+
+    final long numberOfBatches = calculateNumberofBatches(layersToRetain);
+
+    processTrieLogBatches(
+        rootWorldStateStorage,
+        blockchain,
+        chainHeight,
+        lastBlockNumberToRetainTrieLogsFor,
+        numberOfBatches,
+        batchFileNameBase);
+
+    if (rootWorldStateStorage.streamTrieLogKeys(layersToRetain).count() == layersToRetain) {
+      deleteFiles(batchFileNameBase, numberOfBatches);
+      LOG.info("Prune ran successfully. Enjoy some disk space back! \uD83D\uDE80");
+    } else {
+      LOG.error("Prune failed. Re-run the subcommand to load the trie logs from file.");
+    }
+  }
+
+  private static void processTrieLogBatches(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain,
+      final long chainHeight,
+      final long lastBlockNumberToRetainTrieLogsFor,
+      final long numberOfBatches,
+      final String batchFileNameBase) {
+
+    for (long batchNumber = 1; batchNumber <= numberOfBatches; batchNumber++) {
+      saveTrieLogBatches(
+          batchFileNameBase,
+          rootWorldStateStorage,
+          blockchain,
+          chainHeight,
+          lastBlockNumberToRetainTrieLogsFor,
+          batchNumber);
+    }
+
+    LOG.info("Clear trie logs...");
+    rootWorldStateStorage.clearTrieLog();
+
+    for (long batchNumber = 1; batchNumber <= numberOfBatches; batchNumber++) {
+      restoreTrieLogBatches(rootWorldStateStorage, batchNumber, batchFileNameBase);
+    }
+  }
+
+  private static void saveTrieLogBatches(
+      final String batchFileNameBase,
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain,
+      final long chainHeight,
+      final long lastBlockNumberToRetainTrieLogsFor,
+      final long batchNumber) {
+
+    final long firstBlockOfBatch = chainHeight - ((batchNumber - 1) * BATCH_SIZE);
+
+    final long lastBlockOfBatch =
+        Math.max(chainHeight - (batchNumber * BATCH_SIZE), lastBlockNumberToRetainTrieLogsFor);
+
+    final List<Hash> trieLogKeys = new ArrayList<>();
+
+    getTrieLogKeysForBlocks(blockchain, firstBlockOfBatch, lastBlockOfBatch, trieLogKeys);
+
+    LOG.info("Saving trie logs to retain in file (batch {})...", batchNumber);
+
+    try {
+      saveTrieLogsInFile(trieLogKeys, rootWorldStateStorage, batchNumber, batchFileNameBase);
+    } catch (IOException e) {
+      LOG.error("Error saving trie logs to file: {}", e.getMessage());
```

Comment:
it seems that if we have an exception we continue the process, so is it possible we missed some trielog in the restore step in case of failure here

Reply:
going to change this to throw a runtime exception, thanks for catching this